common/unixStatAdapter.go

```diff
@@ -0,0 +1,381 @@
+package common
+
+import (
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"strconv"
+	"time"
+)
+
+const ( // POSIX property metadata
+	POSIXNlinkMeta         = "posix_nlink"
+	POSIXINodeMeta         = "posix_ino"
+	POSIXCTimeMeta         = "posix_ctime"
+	LINUXBTimeMeta         = "linux_btime"
+	POSIXBlockDeviceMeta   = "is_block_dev" // todo: read & use these
+	POSIXCharDeviceMeta    = "is_char_dev"
+	POSIXSocketMeta        = "is_socket"
+	POSIXFIFOMeta          = "is_fifo"
+	POSIXDevMeta           = "posix_dev"
+	POSIXRDevMeta          = "posix_rdev"
+	POSIXATimeMeta         = "posix_atime"
+	POSIXFolderMeta        = "hdi_isfolder" // todo: read & use these
+	POSIXSymlinkMeta       = "is_symlink"
+	POSIXOwnerMeta         = "posix_owner"
+	POSIXGroupMeta         = "posix_group"
+	POSIXModeMeta          = "permissions"
+	POSIXModTimeMeta       = "modtime"
+	LINUXAttributeMeta     = "linux_attribute"
+	LINUXAttributeMaskMeta = "linux_attribute_mask"
+	LINUXStatxMaskMeta     = "linux_statx_mask"
+)
+
+var AllLinuxProperties = []string{
+	POSIXNlinkMeta,
+	POSIXINodeMeta,
+	LINUXBTimeMeta,
+	POSIXBlockDeviceMeta,
+	POSIXCharDeviceMeta,
+	POSIXSocketMeta,
+	POSIXFIFOMeta,
+	POSIXDevMeta,
+	POSIXRDevMeta,
+	POSIXATimeMeta,
+	POSIXFolderMeta,
+	POSIXSymlinkMeta,
+	POSIXOwnerMeta,
+	POSIXGroupMeta,
+	POSIXModeMeta,
+}
+
+//goland:noinspection GoCommentStart
+type UnixStatAdapter interface {
+	Extended() bool // Did this call come from StatX?
+
+	// Statx properties
+	StatxMask() uint32 // Mask determines the availability of all stat/stax properties (except the device ID!). It is only used in statx though.
+	Attribute() uint64 // Attribute is masked by AttributeMask.
+	AttributeMask() uint64
+	BTime() time.Time // BTime may not always be available on every filesystem. It's important to check Mask first!
+	// ==========
+
+	// Base Stat properties
+	NLink() uint64
+	Owner() uint32
+	Group() uint32
+	FileMode() uint32 // Mode may not always be available to check in a Statx call (though it should be, since we requested it.) Best safe than sorry; check Mask!
+	INode() uint64
+	Device() uint64
+	RDevice() uint64 // RDevice is ONLY useful when Mode has S_IFCHR or S_IFBLK; as those determine if the file is a representitive of a block or character device.
+	ATime() time.Time
+	MTime() time.Time
+	CTime() time.Time
+}
+
+type UnixStatContainer struct { // Created for downloads
+	statx bool // Does the call contain extended properties (attributes, birthTime)?
+
+	mask       uint32
+	attributes uint64
+	numLinks   uint64
```

Comment:
I disagree. This struct (and interface) is OS-agnostic, but POSIX specific. If we decide to support some other POSIX-based OS like MacOS or FreeBSD in the future for the POSIX property persistence feature, this could cause problems if we adhere to Linux-specific data typing.

Reply:
I meant those sizes are POSIX specific not Linux specific ("stat()" is a POSIX call and these sizes are not specific to statx()). So any POSIX OS should have the same sizes.